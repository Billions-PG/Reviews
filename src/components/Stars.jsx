import React from 'react';
import PropTypes from 'prop-types';
import FilledStars from '../assets/filled-stars-sm.png';
import EmptyStars from '../assets/empty-stars-sm.png';

/* CSS */
import style from './stars.module.css';

const Stars = ({ rating }) => (
  <span className={style.stars}>
    <img
      className={`${style.stars} ${style.individual}`}
      src={EmptyStars}
      alt=""
      style={{ width: '74x' }}
    />
    <img
      className={`${style.stars} ${style.individual}`}
      src={FilledStars}
      alt=""
      data-testid="rating-star"
      style={{ width: `${rating * (74 / 5)}px` }}
    />
  </span>
);

Stars.propTypes = {
  rating: PropTypes.number.isRequired,
};

export default Stars;
