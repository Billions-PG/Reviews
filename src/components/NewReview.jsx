/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Collapse from 'react-bootstrap/Collapse';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Stars from './Stars';

import style from './newreview.module.css';

const axios = require('axios');

const url = `localhost:${process.env.PORT || 3000}`;

const NewReview = ({ change, prodId }) => {
  const [author, setAuthor] = useState('');
  const [body, setBody] = useState('');
  const [show, toggleShow] = useState(false);
  const [validated, setValidated] = useState(false);
  const [submitted, setSubmit] = useState(false);
  const [xOffset, setXOffset] = useState(0);
  const [tracking, toggleTracking] = useState(true);
  const [purchased, setPurchased] = useState(false);

  useEffect(() => {
    if (submitted) {
      axios.post(`${url}/api/reviews`, {
        author: { name: author }, rating: xOffset, body, purchased, prodId,
      })
        .then(() => change(true));
    }
  }, [submitted]);

  const handleSubmit = (event) => {
    const form = event.currentTarget;
    event.preventDefault();
    if (!form.checkValidity()) {
      event.stopPropagation();
    } else {
      setSubmit(true);
    }
    setValidated(true);
  };

  // eslint-disable-next-line no-underscore-dangle
  const _onMouseMove = (e) => {
    if (tracking) setXOffset((Math.min(e.screenX, -462) + 462 + 73) * (5 / 73));
  };

  const resetForm = () => {
    setAuthor('');
    setBody('');
    setXOffset(0);
    toggleTracking(true);
  };

  return (
    <>
      <div className={style.form}>

        <Button
          data-testid="toggle-form"
          // eslint-disable-next-line no-nested-ternary
          variant={show ? (submitted ? 'success' : 'danger') : 'primary'}
          style={{ borderRadius: `${(show && !submitted) ? '100%' : ''}` }}
          size={show ? 'sm' : 'md'}
          className={style.show}
          type="button"
          onClick={() => {
            toggleShow(!show);
            if (!submitted) resetForm();
          }}
        >
          {/* eslint-disable-next-line no-nested-ternary */}
          {show ? (submitted ? 'done' : 'X') : 'write a review'}
        </Button>
      </div>
      <Collapse className={style.collapsible} in={show}>
        <Form data-testid="new" noValidate validated={validated} onSubmit={handleSubmit}>
          <Form.Group controlId="authorName">
            <Form.Label>Name</Form.Label>
            <Form.Control data-testid="name" disabled={submitted} required type="name" value={author} onChange={(e) => setAuthor(e.target.value)} />
            <Form.Control.Feedback type="invalid">Please provide a name</Form.Control.Feedback>
          </Form.Group>
          <Form.Group>
            <Form.Label>Your rating</Form.Label>
            <span data-testid="mouse-stars" onMouseMove={(e) => _onMouseMove(e)} onClick={() => toggleTracking(false)}>
              <Stars rating={xOffset} />
            </span>
          </Form.Group>
          <Form.Group controlId="reviewBody">
            <Form.Label>Your review</Form.Label>
            <Form.Control data-testid="body" required disabled={submitted} as="textarea" rows={4} value={body} onChange={(e) => setBody(e.target.value)} />
            <Form.Control.Feedback type="invalid">Review body can&#39;t be empty</Form.Control.Feedback>
            <Form.Text>
              <small>
                Your review is subject to terms and services.
                If users find your review offensive, it will be removed.
              </small>
            </Form.Text>
          </Form.Group>
          <Form.Group>
            <Form.Check type="checkbox" label="I bought this product" value={purchased} onChange={() => setPurchased(!purchased)} />
          </Form.Group>
          <Button
            data-testid="submit"
            variant={submitted ? 'secondary' : 'outline-primary'}
            type="submit"
            size="sm"
            disabled={submitted}
          >
            {submitted ? 'review received' : 'submit'}
          </Button>
        </Form>
      </Collapse>
    </>
  );
};

NewReview.propTypes = {
  change: PropTypes.func.isRequired,
  prodId: PropTypes.string.isRequired,
};

export default NewReview;
