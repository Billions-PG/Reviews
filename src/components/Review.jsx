/* eslint-disable no-underscore-dangle */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Image from 'react-bootstrap/Image';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Badge from 'react-bootstrap/Badge';
import Collapse from 'react-bootstrap/Collapse';

import Stars from './Stars';

import style from './review.module.css';

const moment = require('moment');

const Review = ({ review, report }) => {
  const [expanded, toggleExpand] = useState(false);

  const purchased = <small>Purchased item</small>;

  return (
    <div data-testid="review" className={style.review}>
      <Col>
        <Row>
          <Col>
            <Row>
              <Col style={{ maxWidth: 'fit-content' }}>
                <Image className={style.icon} src={review.author.photo} roundedCircle />
              </Col>

              <Col>
                <h5>
                  <p>
                    <small>
                      {review.author.name}
                      <span className={style.spacer} />
                      {moment(review.createdAt).format('MMM D, YYYY')}
                    </small>
                  </p>
                </h5>
                <Row>
                  <Stars rating={review.rating} />
                </Row>
              </Col>
            </Row>
          </Col>
        </Row>
        <Row>
          <Col>
            {review.body.length < 250 ? review.body
              : (
                <div>
                  {`${review.body.substring(0, 250)}`}
                  {!expanded && (
                    <Badge
                      style={{ cursor: 'pointer' }}
                      onClick={() => toggleExpand(true)}
                      pill
                      variant="light"
                    >
                      <b>...</b>
                    </Badge>
                  )}
                  <Collapse in={expanded}>
                    <span>
                      {review.body.substring(250)}
                    </span>

                  </Collapse>
                </div>
              )}
            <br />
            {(review.purchased || review.photo) && purchased}
          </Col>
          <Col md="auto">

            <Row>{review.photo ? <Image src={review.photo} rounded /> : <></>}</Row>

          </Col>
        </Row>
        <Row className={style.report}>
          <small data-testid="report" className={style.report} onClick={() => report(review._id)}>report</small>
        </Row>
      </Col>
    </div>
  );
};

Review.propTypes = {
  review: PropTypes.shape({
    author: PropTypes.shape({
      name: PropTypes.string.isRequired,
      photo: PropTypes.string,
    }),
    rating: PropTypes.number.isRequired,
    purchased: PropTypes.bool.isRequired,
    body: PropTypes.string.isRequired,
    photo: PropTypes.string,
    createdAt: PropTypes.string,
    _id: PropTypes.string.isRequired,
  }).isRequired,
  report: PropTypes.func.isRequired,
};

export default Review;
