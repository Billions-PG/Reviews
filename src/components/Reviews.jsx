/* eslint-disable no-underscore-dangle */
import React from 'react';
import PropTypes from 'prop-types';
import Container from 'react-bootstrap/Container';

import FilledStars from '../assets/filled-stars.png';
import EmptyStars from '../assets/empty-stars.png';
import Review from './Review';

/* CSS */
import style from './stars.module.css';

const Controller = require('../../controllers');

const Reviews = (props) => {
  const [reviews, setReviews] = React.useState([]);
  const [avgRating, setAvgRating] = React.useState(5);
  React.useEffect(() => {
    (async () => {
      try {
        const res = await Controller.getReview(props.prodId);
        setReviews(res);
      } catch (err) {
        setReviews([]);
      }
    })();
  }, [props]);

  React.useEffect(() => {
    if (!reviews.length) { return; }
    const avg = reviews.reduce((m, i) => m + i.rating, 0) / reviews.length;
    setAvgRating(avg);
  }, [reviews]);

  return (
    <>
      {!reviews.length ? 'Loading...'
        : (
          <Container>
            <h3 data-testid="review-count">
              {`${reviews.length} review${reviews.length > 1 ? 's' : ''}`}
              <div className={style.stars}>
                <img
                  className={style.stars}
                  src={EmptyStars}
                  alt=""
                  style={{ width: '244px' }}
                />
                <img
                  className={style.stars}
                  src={FilledStars}
                  alt=""
                  data-testid="rating-stars"
                  style={{ width: `${avgRating * (244 / 5)}px` }}
                />
              </div>
            </h3>
            <hr />
            {reviews.map((review) => (
              <>
                <Review review={review} />
                <hr />
              </>
            ))}
            {/* <Row>{photoArray.length ? photos : <></>}</Row> */}
          </Container>
        )}
    </>
  );
};

Reviews.propTypes = {
  prodId: PropTypes.string.isRequired,
};

export default Reviews;
