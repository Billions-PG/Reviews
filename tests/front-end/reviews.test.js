/* eslint-disable react/jsx-filename-extension */
/**
 * @test-environment jsdom
 */
import 'jsdom-global/register';
import '@testing-library/jest-dom';
import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { useParams } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import Reviews from '../../src/components/Reviews';
import Review from '../../src/components/Review';
import Stars from '../../src/components/Stars';
import NewReview from '../../src/components/NewReview';

const axios = require('axios');

jest.mock('axios');
axios.post.mockResolvedValue('posted');

const Controller = require('../../controllers');

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  clear: jest.fn(),
};
global.localStorage = localStorageMock;

// HELPER FUNCTION FOR ROUTER
// https://medium.com/@aarling/mocking-a-react-router-match-object-in-your-component-tests-fa95904dcc55

describe('<Review />', () => {
  let mockReview;
  beforeAll(() => {
    mockReview = {
      author: {
        name: 'Jon',
      },
      rating: 5,
      purchased: false,
      body: 'Great product!',
      photo: 'sampleWebpage.com/100.jpg',
      _id: '1',
    };
  });

  test('creates a review from the author name, rating, purchased, and body', () => {
    const { queryByRole } = render(<Review report={jest.fn()} review={mockReview} />);
    expect(queryByRole('heading')).toBeTruthy();
  });
  test('clicking "report" calls a delete with the reviews _id', async () => {
    const mockReport = jest.fn();
    const { getByTestId } = render(<Review report={mockReport} review={mockReview} />);
    const report = getByTestId('report');
    userEvent.click(report);
    await waitFor(() => expect(mockReport).toHaveBeenCalledWith('1'));
  });
});

describe('<Reviews />', () => {
  let controllerSpy;
  let mockReview;
  beforeEach(() => {
    // useParams is mocked at ../../__mocks__/react-router-dom.js
    useParams.mockReturnValue({ prodId: '1' });
    controllerSpy = jest.spyOn(Controller, 'getReview');
    mockReview = {
      author: {
        name: 'Jon',
      },
      rating: 5,
      purchased: false,
      body: 'Great product!',
      photo: 'sampleWebpage.com/100.jpg',
    };
    controllerSpy.mockResolvedValue([
      {
        ...mockReview, _id: '1', prodId: '1', rating: 5,
      },
      {
        ...mockReview, _id: '2', prodId: '1', rating: 0,
      }]);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  test('mocks exist', () => {
    expect(jest.isMockFunction(useParams)).toBeTruthy();
  });
  test('calls Contoller.getReview on initialization', async () => {
    render(<Reviews />);
    await waitFor(() => expect(controllerSpy).toHaveBeenCalled());
  });
  test('renders a review for each product returned by Controller.getReview', async () => {
    const { findByTestId } = render(<Reviews />);
    const reviews = await findByTestId('review-count');
    expect(reviews).toHaveTextContent('2 reviews');
  });
  test('renders stars proportional to the average rating', async () => {
    const { findByTestId } = render(<Reviews prodId="1" />);
    const stars = await findByTestId('rating-stars');
    expect(stars).toHaveStyle({ width: '122px' });
  });
});

describe('<Stars />', () => {
  test('filled in stars are proportional in width to the rating', () => {
    const { getByTestId } = render(<Stars rating={4} />);
    const star = getByTestId('rating-star');
    expect(star).toHaveStyle({ width: '59.2px' });
  });
  test('uses CSS modules', () => {
    const { getByTestId } = render(<Stars rating={4} />);
    const star = getByTestId('rating-star');
    expect(star).not.toHaveClass('star');
    expect(star).not.toHaveClass('individual');
  });
});

describe('<NewReview />', () => {
  let changeMock;
  let prodId;
  let component;
  beforeAll(() => {
    changeMock = jest.fn();
    prodId = '50';
  });
  beforeEach(() => {
    component = render(<NewReview change={changeMock} prodId={prodId} />);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  test('it has a form with a required name and body section', () => {
    const { getByTestId } = component;
    const form = getByTestId('new');
    const nameInput = getByTestId('name');
    const bodyInput = getByTestId('body');
    expect(form).toBeInTheDocument();
    expect(bodyInput).toBeRequired();
    expect(nameInput).toBeRequired();
  });
  test('does not submit if form is not valid', async () => {
    const { getByTestId } = component;
    const form = getByTestId('new');
    const submitBtn = getByTestId('submit');
    expect(form).not.toBeValid();
    userEvent.click(submitBtn);
    await waitFor(() => expect(axios.post).not.toHaveBeenCalled());
  });
  test('submits form if validation is complete', async () => {
    const { getByTestId } = component;
    const form = getByTestId('new');
    const nameInput = getByTestId('name');
    const bodyInput = getByTestId('body');
    const submitBtn = getByTestId('submit');
    userEvent.type(nameInput, 'a name');
    userEvent.type(bodyInput, 'some input here');
    expect(nameInput).toHaveValue('a name');
    expect(bodyInput).toHaveValue('some input here');
    expect(form).toBeValid();
    userEvent.click(submitBtn);
    await waitFor(() => expect(axios.post).toHaveBeenCalled());
  });
  test('moving the mouse over the stars changes the width', () => {
    const { asFragment, getByTestId } = component;
    const starHolder = getByTestId('mouse-stars');
    userEvent.hover(starHolder);
    expect(asFragment()).toMatchSnapshot();
  });
  test('toggle the form resets the state values if not submitted', () => {
    const { asFragment, getByTestId } = component;
    const nameInput = getByTestId('name');
    const bodyInput = getByTestId('body');
    userEvent.type(nameInput, 'a name');
    userEvent.type(bodyInput, 'some input here');
    const toggle = getByTestId('toggle-form');
    userEvent.dblClick(toggle);
    expect(asFragment()).toMatchSnapshot();
  });
});
