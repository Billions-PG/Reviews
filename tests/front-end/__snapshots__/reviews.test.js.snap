// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewReview /> moving the mouse over the stars changes the width 1`] = `
<DocumentFragment>
  <div>
    <button
      class="btn btn-primary btn-md"
      data-testid="toggle-form"
      type="button"
    >
      write a review
    </button>
  </div>
  <form
    class="collapse"
    data-testid="new"
    novalidate=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="authorName"
      >
        Name
      </label>
      <input
        class="form-control"
        data-testid="name"
        id="authorName"
        required=""
        type="name"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Please provide a name
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        Your rating
      </label>
      <span
        data-testid="mouse-stars"
      >
        <span>
          <img
            alt=""
            class="undefined undefined"
            src=""
          />
          <img
            alt=""
            class="undefined undefined"
            data-testid="rating-star"
            src=""
            style="width: 74px;"
          />
        </span>
      </span>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="reviewBody"
      >
        Your review
      </label>
      <textarea
        class="form-control"
        data-testid="body"
        id="reviewBody"
        required=""
        rows="4"
      />
      <div
        class="invalid-feedback"
      >
        Review body can't be empty
      </div>
      <small
        class="form-text"
      >
        <small>
          Your review is subject to terms and services. If users find your review offensive, it will be removed.
        </small>
      </small>
    </div>
    <div
      class="form-group"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          type="checkbox"
          value="false"
        />
        <label
          class="form-check-label"
          title=""
        >
          I bought this product
        </label>
      </div>
    </div>
    <button
      class="btn btn-outline-primary btn-sm"
      data-testid="submit"
      type="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<NewReview /> toggle the form resets the state values if not submitted 1`] = `
<DocumentFragment>
  <div>
    <button
      class="btn btn-primary btn-md"
      data-testid="toggle-form"
      style=""
      type="button"
    >
      write a review
    </button>
  </div>
  <form
    class="collapsing"
    data-testid="new"
    novalidate=""
    style=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="authorName"
      >
        Name
      </label>
      <input
        class="form-control"
        data-testid="name"
        id="authorName"
        required=""
        type="name"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Please provide a name
      </div>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
      >
        Your rating
      </label>
      <span
        data-testid="mouse-stars"
      >
        <span>
          <img
            alt=""
            class="undefined undefined"
            src=""
          />
          <img
            alt=""
            class="undefined undefined"
            data-testid="rating-star"
            src=""
            style="width: 0px;"
          />
        </span>
      </span>
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="reviewBody"
      >
        Your review
      </label>
      <textarea
        class="form-control"
        data-testid="body"
        id="reviewBody"
        required=""
        rows="4"
      />
      <div
        class="invalid-feedback"
      >
        Review body can't be empty
      </div>
      <small
        class="form-text"
      >
        <small>
          Your review is subject to terms and services. If users find your review offensive, it will be removed.
        </small>
      </small>
    </div>
    <div
      class="form-group"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          type="checkbox"
          value="false"
        />
        <label
          class="form-check-label"
          title=""
        >
          I bought this product
        </label>
      </div>
    </div>
    <button
      class="btn btn-outline-primary btn-sm"
      data-testid="submit"
      type="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;
